### PR TITLE
Make discriminated unions trimmable

### PR DESCRIPTION
**Pull request by KevinRansom**
- [x] Ad trimable test
- [x] Trimmable discriminated unions
- [x] Update baselines

### CODE REVIEW

File: tests/projects/SelfContained_Trimming_Test/check.ps1 [hunk 1]
```diff
@@ -14,7 +14,7 @@ if (-not ($output -eq $expected))
 }
 
 # Checking that FSharp.Core binary is of expected size (needs adjustments if test is updated).
-$expected_len = 251904 # In bytes
+$expected_len = 264704 # In bytes
```
Comment (T-Gro) [inline tests/projects/SelfContained_Trimming_Test/check.ps1:17]: This time the increase of trimmed Fsharp.Core is quite big.
Do we know what contributes the most to the code size here?

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/SteppingMatch/SteppingMatch07.fs.il.netcore.release.bsl [hunk 2]
```diff
@@ -91,8 +91,9 @@
     .method assembly specialname rtspecialname 
             instance void  .ctor(int32 _tag) cil managed
     {
-      .custom instance void [runtime]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = ( 01 00 00 00 )
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/EmittedIL/SteppingMatch/SteppingMatch07.fs.il.netcore.release.bsl:94]: Why is baseline updating dropping the CompilerGeneratedAttribute and DebuggerNonUserCodeAttribute attributes in most of the files?

I would have expected only an addition of DynamicDependencyAttribute.

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/GenericComparison/Compare07.fsx.il.net472.bsl [hunk 3]
```diff
@@ -0,0 +1,765 @@
+
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/EmittedIL/GenericComparison/Compare07.fsx.il.net472.bsl:1]: I can see a lot of .bsl added with this PR.

E.g. for Compare07, we have:

il.bsl
ill.netcore.bsl
il.net472.bsl

What is the first one (without specified platform) supposed to indicate?

Comment (KevinRansom) [reply tests/FSharp.Compiler.ComponentTests/EmittedIL/GenericComparison/Compare07.fsx.il.net472.bsl:1]: that is a bug, really good eyes.  the issue is the ill.netcore should be il.netcore.bsl, and il.bsl shouldn't e xist.

File: tests/projects/SelfContained_Trimming_Test/check.ps1 (hunk 1 above)
Comment (KevinRansom) [reply tests/projects/SelfContained_Trimming_Test/check.ps1:17]: @t-gro we use discriminated unions quite a bunch in FSharp.Core.  This change will have caused their Properties and Nested types to be kept.

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/SteppingMatch/SteppingMatch07.fs.il.netcore.release.bsl (hunk 2 above)
Comment (KevinRansom) [reply tests/FSharp.Compiler.ComponentTests/EmittedIL/SteppingMatch/SteppingMatch07.fs.il.netcore.release.bsl:94]: Yes we shouldn't be dropping the compiler generated attributes.  I will try to figure out why that is so.

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/GenericComparison/Compare07.fsx.il.net472.bsl (hunk 3 above)
Comment (KevinRansom) [reply tests/FSharp.Compiler.ComponentTests/EmittedIL/GenericComparison/Compare07.fsx.il.net472.bsl:1]: fixed

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/SteppingMatch/SteppingMatch07.fs.il.netcore.release.bsl (hunk 2 above)
Comment (KevinRansom) [reply tests/FSharp.Compiler.ComponentTests/EmittedIL/SteppingMatch/SteppingMatch07.fs.il.netcore.release.bsl:94]: Fixed

**Review by T-Gro — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]

**Pull request merged**